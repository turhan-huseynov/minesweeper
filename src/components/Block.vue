<template>
    <div class="block unselectable" :class="{ mine: isMine, highlight: shouldHighlight, flipped: isFlipped }">
        <div class="flip-block" :class="{ flipped: isFlipped }" @click="clickFunc">
            <div class="flip-block-inner">
                <div class="flip-block-front"></div>
                <div class="flip-block-back">
                    <img v-if="isMine" alt="mine" height="30" width="30" src="@/assets/mine.jpg" />
                    <div v-else>{{ blockText || "" }}</div>
                </div>
            </div>
        </div>
    </div>
</template>

<script setup>
import { defineProps } from "vue";
/**
 * isMine: decides if block will be a mine
 * shouldHighlight: decides if background of block will be darker. used for giving chessboard look
 * isFlipped: decides if block clicked
 * clickFunc: function to trigger the block click
 */
defineProps({
    isMine: { type: Boolean, default: false },
    shouldHighlight: { type: Boolean, default: false },
    isFlipped: { type: Boolean, default: false },
    blockText: { type: [String, Number], default: "" },
    clickFunc: {
        type: Function,
        default: () => console.log("Click function not implemented"),
    },
});
</script>

<style lang="scss" scoped>
.block {
    cursor: pointer;

    &.highlight .flip-block .flip-block-inner .flip-block-front {
        background-color: rgb(119, 189, 119);
    }

    &.mine .flip-block .flip-block-inner .flip-block-back {
        background-color: rgb(255, 70, 70);
        color: white;
    }

    &:hover {
        filter: brightness(0.8);
    }

    .flip-block {
        background-color: transparent;
        width: 40px;
        height: 40px;
        perspective: 1000px;

        .flip-block-inner {
            position: relative;
            width: 100%;
            height: 100%;
            transition: transform 0.8s;

            .flip-block-front,
            .flip-block-back {
                position: absolute;
                width: 100%;
                height: 100%;
                display: flex;
                justify-content: center;
                align-items: center;
                border-radius: 2px;
            }

            .flip-block-front {
                background-color: rgb(135, 221, 135);

                color: black;
                z-index: 1;
            }

            .flip-block-back {
                background-color: lightgray;
                color: black;
            }
        }

        &.flipped .flip-block-inner .flip-block-back {
            z-index: 5;
        }
    }
}
</style>
