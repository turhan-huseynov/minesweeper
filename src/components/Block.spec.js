import { render, fireEvent, waitFor } from "@testing-library/vue";
import { expect, describe, it, afterEach } from "vitest";
import Block from "./Block.vue";

describe("Block component", () => {
    let wrapper;

    afterEach(() => {
        if (wrapper) {
            wrapper.unmount();
        }
    });

    it("should match the snapshot", () => {
        wrapper = render(Block, { props: { isFlipped: true, shouldHighlight: true, blockText: 5 } });

        expect(wrapper.container).toMatchSnapshot();
    });

    it("should render a number block", () => {
        wrapper = render(Block, { props: { blockText: 5 } });
        const numberElem = wrapper.getByText(5);
        expect(numberElem).toBeTruthy();
    });

    it("should render a mine", () => {
        wrapper = render(Block, { props: { isMine: true, blockText: "Mine" } });
        const button = wrapper.getByText("Mine");
        expect(button).toBeTruthy();
    });

    it("should call the prop function", async () => {
        let isCalled = false;
        const fn = () => {
            isCalled = true;
        };
        wrapper = render(Block, { props: { clickFunc: fn, blockText: "Test" } });
        const blockElem = wrapper.getByText("Test");

        await fireEvent.click(blockElem);
        await waitFor(() => {
            expect(isCalled).toBeTruthy();
        });
    });

    it("should flip block after click", async () => {
        wrapper = render(Block, { props: { blockText: "Deneme", isFlipped: true } });
        const flippedElem = wrapper.container.getElementsByClassName("flipped");
        expect(flippedElem.length).toBe(2);
    });

    // TODO: Add test for Mine click
});
