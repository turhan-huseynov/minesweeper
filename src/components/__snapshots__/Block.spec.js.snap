// Vitest Snapshot v1

exports[`Block component > should match the snapshot 1`] = `
<div>
  <div
    class="block unselectable highlight flipped"
    data-v-22c6a20d=""
  >
    <div
      class="flip-block flipped"
      data-v-22c6a20d=""
    >
      <div
        class="flip-block-inner"
        data-v-22c6a20d=""
      >
        <div
          class="flip-block-front"
          data-v-22c6a20d=""
        />
        <div
          class="flip-block-back"
          data-v-22c6a20d=""
        >
          5
        </div>
      </div>
    </div>
  </div>
</div>
`;
